crates/resolver/src/name_server_pool.rs

```diff
@@ -93,6 +115,35 @@ impl<P: ConnectionProvider> DnsHandle for NameServerPool<P> {
     type Response = Pin<Box<dyn Stream<Item = Result<DnsResponse, ProtoError>> + Send>>;
     type Runtime = P::RuntimeProvider;
 
+    fn lookup(&self, query: Query, mut options: DnsRequestOptions) -> Self::Response {
+        let query_copy = query.clone();
+        options.case_randomization = self.state.cx.options.case_randomization;
+        // block concurrent requests
+        let lookup = self
+            .active_requests
+            .lock()
+            .entry(query.clone())
+            .or_insert_with(move || {
+                info!("querying {query_copy}");
+
+                // convert the lookup into a shared future
+                let lookup = self
+                    .send(DnsRequest::from_query(query_copy, options))
+                    .into_future()
+                    .map(|(next, _)| next)
+                    .boxed()
+                    .shared();
+
+                SharedLookup(lookup)
+            })
+            .clone();
+
+        // remove the concurrent request marker
+        self.active_requests.lock().remove(&query);
+
+        Box::pin(once(lookup))
```

Comment:
Comparing to the previous code, this is missing an await on the lookup future. Thus, this will store an entry into `active_requests`, drop the lock, retake it, and remove the entry. We'll need to instead return something like a pinned boxed async block, which awaits on the lookup, removes the active requests entry, and then returns a stream that returns the result immediately.